packages/react-transition-presence/src/_utils/childrenAreEqual.ts

```diff
@@ -0,0 +1,34 @@
+import { type ReactElement, type ReactFragment } from 'react';
+import { shallowCompare } from './shallowCompare.js';
+
+export function childrenAreEqual(
+  previousChildren: ReactElement | ReactFragment | null,
+  nextChildren: ReactElement | ReactFragment | null,
+): boolean {
+  if (previousChildren === nextChildren) {
+    return true;
+  }
+
+  if (
+    (previousChildren === null && nextChildren !== null) ||
+    (previousChildren !== null && nextChildren === null)
+  ) {
+    return false;
+  }
+
+  if (
+    ('type' in previousChildren! && previousChildren.type) === // eslint-disable-line @typescript-eslint/no-non-null-assertion
+    ('type' in nextChildren! && nextChildren.type) // eslint-disable-line @typescript-eslint/no-non-null-assertion
+  ) {
+    if (
+      'props' in previousChildren! && // eslint-disable-line @typescript-eslint/no-non-null-assertion
+      'props' in nextChildren! // eslint-disable-line @typescript-eslint/no-non-null-assertion
+    ) {
+      return shallowCompare(previousChildren.props, nextChildren.props);
```

Comment:
Does this also work for nested children (i.e. multiple wrappers)?

Since `nextChildren.props.children` might fail the shallow check.

Also, would this break of any passed prop doesn't use memo/callback/etc? I don't think we should rely on that always being the case.

Reply:
It's currently not comparing any props of children right now, it'll only call the before unmount callbacks when a new component instance is created from the children. A `key` **must** be used if a transition should start when a nested wrapper is changed/updated.